ddtrace/internal/utils/http.py

```diff
@@ -38,6 +40,60 @@ def strip_query_string(url):
     return h + fs + f
 
 
+def redact_query_string(query_string, query_string_obfuscation_pattern):
+    # type: (str, Optional[re.Pattern]) -> Union[bytes, str]
+    if query_string_obfuscation_pattern is None:
+        return query_string
+
+    bytes_query = query_string if isinstance(query_string, bytes) else query_string.encode("utf-8")
+    return query_string_obfuscation_pattern.sub(b"<redacted>", bytes_query)
+
+
+def redact_url(url, query_string_obfuscation_pattern, query_string=None):
+    # type: (str, re.Pattern, Optional[str]) -> Union[str,bytes]
+
+    # Avoid further processing if obfuscation is disabled
+    if query_string_obfuscation_pattern is None:
+        return url
+
+    parts = compat.parse.urlparse(url)
+    redacted_query = None
+
+    if query_string:
+        redacted_query = redact_query_string(query_string, query_string_obfuscation_pattern)
+    elif parts.query:
+        redacted_query = redact_query_string(parts.query, query_string_obfuscation_pattern)
+
+    if redacted_query is not None and len(parts) >= 5:
+        redacted_parts = parts[:4] + (redacted_query,) + parts[5:]  # type: Tuple[Union[str, bytes], ...]
+        bytes_redacted_parts = tuple(x if isinstance(x, bytes) else x.encode("utf-8") for x in redacted_parts)
+        return urlunsplit(bytes_redacted_parts, url)
+
+    # If no obfuscation is performed, return original url
+    return url
+
+
+def urlunsplit(components, original_url):
+    # type: (Tuple[bytes, ...], str) -> bytes
+    """
+    Adaptation from urlunsplit and urlunparse, using bytes components
+    """
+    scheme, netloc, url, params, query, fragment = components
+    if params:
```

Comment:
Can we use `urlunsplit` and/or  `urlunparse` and then convert the url to a byte string? This might be easier to maintain. Recently we fixed a bug in our request integration which was introduced by custom url parser

Reply:
I tried that and ran into some problems, I agree that it would be easier to maintain. I will give it a try again

Reply:
The problem is, to support the obfuscation of byte query strings, we need to convert to bytes prior to passing it to the regexp, so the output of the obfuscation is always bytes.